File: docs/archive/packages/python-sdk/examples/basic_usage.py
"""
Basic Usage Example

Shows how an AI agent queries organizational intent before taking action.
"""

import sys
from pathlib import Path

# Add package to path
sys.path.insert(0, str(Path(__file__).parent.parent))

from narrative_sdk import IntentClient

# Initialize client (path from repository root)
DB_PATH = str(Path(__file__).parent.parent.parent.parent / '.narrative' / 'intent.db')
client = IntentClient(DB_PATH)

# Example 1: Query intent before writing authentication code
print("=" * 60)
print("Example 1: Writing Authentication Code")
print("=" * 60)

result = client.query_intent(
    operation='writing authentication code',
    context={'tags': ['security', 'authentication']}
)

print(f"\n📋 Intent Chain ({len(result['intentChain'])} units):")
for item in result['intentChain']:
    print(f"  [{item['type']}] {item['assertion'][:70]}...")

print(f"\n🔒 Code Constraints:")
if 'code' in result['constraints']:
    code = result['constraints']['code']

    if 'required_patterns' in code:
        print(f"\n  Required Patterns ({len(code['required_patterns'])}):")
        for pattern in code['required_patterns'][:5]:
            print(f"    • {pattern}")
        if len(code['required_patterns']) > 5:
            print(f"    ... and {len(code['required_patterns']) - 5} more")

    if 'forbidden_patterns' in code:
        print(f"\n  Forbidden Patterns ({len(code['forbidden_patterns'])}):")
        for pattern in code['forbidden_patterns'][:5]:
            print(f"    ✗ {pattern}")
        if len(code['forbidden_patterns']) > 5:
            print(f"    ... and {len(code['forbidden_patterns']) - 5} more")

    if 'required_libraries' in code:
        print(f"\n  Required Libraries:")
        for lib in code['required_libraries']:
            print(f"    • {lib}")

if result['evidenceRequired']:
    print(f"\n📊 Evidence Required:")
    for evidence in result['evidenceRequired'][:5]:
        print(f"  • {evidence}")

# Example 2: Query intent before creating API endpoint
print("\n\n" + "=" * 60)
print("Example 2: Creating API Endpoint")
print("=" * 60)

result = client.query_intent(
    operation='creating API endpoint',
    context={'tags': ['api', 'backend']}
)

print(f"\n📋 Intent Chain ({len(result['intentChain'])} units):")
for item in result['intentChain']:
    print(f"  [{item['type']}] {item['assertion'][:70]}...")

if 'code' in result['constraints']:
    code = result['constraints']['code']
    print(f"\n🔒 Code Constraints:")
    if 'required_patterns' in code:
        print(f"  Required: {', '.join(code['required_patterns'][:5])}")

# Example 3: Query intent before writing marketing content
print("\n\n" + "=" * 60)
print("Example 3: Writing Marketing Content")
print("=" * 60)

result = client.query_intent(
    operation='writing marketing content',
    context={'tags': ['communication', 'marketing']}
)

print(f"\n📋 Intent Chain ({len(result['intentChain'])} units):")
for item in result['intentChain']:
    print(f"  [{item['type']}] {item['assertion'][:70]}...")

if 'content' in result['constraints']:
    content = result['constraints']['content']
    print(f"\n📝 Content Constraints:")
    if 'required_themes' in content:
        print(f"  Required Themes: {', '.join(content['required_themes'])}")
    if 'forbidden_themes' in content:
        print(f"  Forbidden Themes: {', '.join(content['forbidden_themes'])}")
    if 'tone' in content:
        print(f"  Tone: {content['tone']}")

# Example 4: Get graph statistics
print("\n\n" + "=" * 60)
print("Example 4: Graph Statistics")
print("=" * 60)

stats = client.get_stats()
print(f"\n📊 Intent Graph Stats:")
print(f"  Total units: {stats['total']}")
print(f"\n  By Type:")
for type_name, count in sorted(stats['byType'].items()):
    print(f"    {type_name}: {count}")
print(f"\n  By Validation State:")
for state, count in sorted(stats['byValidation'].items()):
    print(f"    {state}: {count}")

# Clean up
client.close()

print("\n✅ Examples complete!")

File: docs/archive/examples/basic-usage.ts
/**
 * Basic Usage Example
 *
 * Demonstrates how an autonomous agent queries organizational intent
 * before performing an operation.
 */

import { IntentClient, IntentUnit } from '@narrative/sdk';

async function main() {
  // Initialize the client
  const client = new IntentClient('.narrative/intent.db');

  // ============================================================================
  // SETUP: Seed the graph with example organizational intent
  // ============================================================================

  // Core Story: Strategic intent from CEO
  const coreStory: IntentUnit = {
    id: 'core_story_healthcare_focus',
    type: 'core_story',
    assertion: 'We are building healthcare infrastructure that protects patient data',
    intent: {
      objective: 'Become the most trusted healthcare infrastructure provider',
      constraints: {
        code: {
          required_patterns: ['audit_logging', 'encryption_at_rest'],
          forbidden_patterns: ['console.log', 'localStorage'],
        },
        content: {
          required_themes: ['security', 'compliance', 'trust'],
          forbidden_themes: ['fast_iteration', 'move_fast_break_things'],
          tone: 'professional',
          target_audience: 'healthcare_decision_makers',
        },
      },
      evidence_required: ['SOC2_certification', 'HIPAA_compliance_audit'],
    },
    dependencies: [],
    validationState: 'ALIGNED',
    confidence: 1.0,
    metadata: {
      created_by: 'ceo',
      tags: ['healthcare', 'security', 'compliance'],
    },
  };

  // Operational: How engineering executes
  const operational: IntentUnit = {
    id: 'operational_auth_implementation',
    type: 'operational',
    assertion: 'All authentication must use OAuth 2.0 with refresh tokens',
    intent: {
      objective: 'Implement secure, auditable authentication',
      constraints: {
        code: {
          required_libraries: ['@aws-sdk/client-kms', 'jsonwebtoken'],
          required_patterns: ['jwt_signing', 'token_rotation'],
          forbidden_patterns: ['basic_auth', 'password_storage'],
        },
        validation_rules: [
          {
            type: 'ast_pattern',
            check: 'no-localStorage-for-tokens',
            error_message: 'Tokens must not be stored in localStorage',
            suggestion: 'Use httpOnly cookies or secure session storage',
          },
          {
            type: 'regex',
            check: 'audit\\(.*auth',
            error_message: 'Authentication events must be audited',
            suggestion: 'Add audit() call after authentication',
          },
        ],
      },
      evidence_required: ['auth_test_coverage_90_percent'],
    },
    dependencies: ['core_story_healthcare_focus'],
    validationState: 'ALIGNED',
    confidence: 0.95,
    metadata: {
      created_by: 'cto',
      tags: ['authentication', 'security', 'oauth'],
    },
  };

  // Create the intent units
  await client.createUnit(coreStory);
  await client.createUnit(operational);

  console.log('✅ Intent graph seeded with 2 units\n');

  // ============================================================================
  // AGENT USE CASE: Coding agent about to write authentication code
  // ============================================================================

  console.log('🤖 Agent Query: Writing authentication code...\n');

  const intentResponse = await client.queryIntent({
    operation: 'writing authentication code',
    context: {
      file_path: 'src/auth/login.ts',
      tags: ['authentication'],
    },
  });

  console.log('📋 Intent Chain:');
  intentResponse.intentChain.forEach(intent => {
    console.log(`  [${intent.type}] ${intent.assertion}`);
  });

  console.log('\n🔒 Code Constraints:');
  console.log('  Required Patterns:', intentResponse.constraints.code?.required_patterns);
  console.log('  Forbidden Patterns:', intentResponse.constraints.code?.forbidden_patterns);
  console.log('  Required Libraries:', intentResponse.constraints.code?.required_libraries);

  console.log('\n📝 Content Constraints:');
  console.log('  Tone:', intentResponse.constraints.content?.tone);
  console.log('  Required Themes:', intentResponse.constraints.content?.required_themes);

  console.log('\n✅ Validation Rules:');
  intentResponse.validationRules.forEach(rule => {
    console.log(`  - ${rule.check}: ${rule.error_message}`);
  });

  console.log('\n📊 Evidence Required:');
  intentResponse.evidenceRequired?.forEach(evidence => {
    console.log(`  - ${evidence}`);
  });

  // ============================================================================
  // AGENT DECISION: Use constraints to guide code generation
  // ============================================================================

  console.log('\n\n🎯 Agent Decision Logic:\n');

  // Check if localStorage is forbidden
  if (intentResponse.constraints.code?.forbidden_patterns?.includes('localStorage')) {
    console.log('❌ Cannot use localStorage for tokens (forbidden pattern)');
    console.log('✅ Will use httpOnly cookies instead');
  }

  // Check required libraries
  if (intentResponse.constraints.code?.required_libraries?.includes('jsonwebtoken')) {
    console.log('✅ Will import jsonwebtoken library');
  }

  // Check required patterns
  if (intentResponse.constraints.code?.required_patterns?.includes('audit_logging')) {
    console.log('✅ Will add audit logging to authentication flow');
  }

  // ============================================================================
  // PROPAGATION: What if we change the core story?
  // ============================================================================

  console.log('\n\n🔄 Propagation Analysis:\n');

  const impact = await client.getPropagationImpact('core_story_healthcare_focus');
  console.log(`Changing core_story_healthcare_focus would affect ${impact.length} units:`);
  impact.forEach(unit => {
    console.log(`  - ${unit.id} (${unit.type})`);
  });

  // ============================================================================
  // STATS
  // ============================================================================

  console.log('\n\n📊 Graph Statistics:\n');
  const stats = await client.getStats();
  console.log('  Total Units:', stats.total);
  console.log('  By Type:', stats.byType);
  console.log('  By Validation:', stats.byValidation);

  // Clean up
  client.close();
}

// Run the example
main().catch(console.error);
